dlt/destinations/impl/clickhouse/clickhouse.py

```diff
@@ -0,0 +1,402 @@
+import os
+import re
+from copy import deepcopy
+from typing import ClassVar, Optional, Dict, List, Sequence, cast, Tuple
+from urllib.parse import urlparse
+
+import clickhouse_connect
+from clickhouse_connect.driver.tools import insert_file
+
+import dlt
+from dlt import config
+from dlt.common.configuration.specs import (
+    CredentialsConfiguration,
+    AzureCredentialsWithoutDefaults,
+    GcpCredentials,
+    AwsCredentialsWithoutDefaults,
+)
+from dlt.common.destination import DestinationCapabilitiesContext
+from dlt.common.destination.reference import (
+    SupportsStagingDestination,
+    TLoadJobState,
+    FollowupJob,
+    LoadJob,
+    NewLoadJob,
+    StorageSchemaInfo,
+)
+from dlt.common.schema import Schema, TColumnSchema
+from dlt.common.schema.typing import (
+    TTableFormat,
+    TTableSchema,
+    TColumnHint,
+    TColumnType,
+    TTableSchemaColumns,
+    TColumnSchemaBase,
+)
+from dlt.common.storages import FileStorage
+from dlt.destinations.exceptions import LoadJobTerminalException
+from dlt.destinations.impl.clickhouse import capabilities
+from dlt.destinations.impl.clickhouse.clickhouse_adapter import (
+    TTableEngineType,
+    TABLE_ENGINE_TYPE_HINT,
+)
+from dlt.destinations.impl.clickhouse.configuration import (
+    ClickHouseClientConfiguration,
+)
+from dlt.destinations.impl.clickhouse.sql_client import ClickHouseSqlClient
+from dlt.destinations.impl.clickhouse.utils import (
+    convert_storage_to_http_scheme,
+    FILE_FORMAT_TO_TABLE_FUNCTION_MAPPING,
+    SUPPORTED_FILE_FORMATS,
+)
+from dlt.destinations.job_client_impl import (
+    SqlJobClientBase,
+    SqlJobClientWithStaging,
+)
+from dlt.destinations.job_impl import NewReferenceJob, EmptyLoadJob
+from dlt.destinations.sql_jobs import SqlMergeJob
+from dlt.destinations.type_mapping import TypeMapper
+
+
+HINT_TO_CLICKHOUSE_ATTR: Dict[TColumnHint, str] = {
+    "primary_key": "PRIMARY KEY",
+    "unique": "",  # No unique constraints available in ClickHouse.
+    "foreign_key": "",  # No foreign key constraints support in ClickHouse.
+}
+
+TABLE_ENGINE_TYPE_TO_CLICKHOUSE_ATTR: Dict[TTableEngineType, str] = {
+    "merge_tree": "MergeTree",
+    "replicated_merge_tree": "ReplicatedMergeTree",
+}
+
+
+class ClickHouseTypeMapper(TypeMapper):
+    sct_to_unbound_dbt = {
+        "complex": "JSON",
```

Comment:
do we have the same problem when `jsonl` and `parquet` are used? the difference
- jsonl stores complex as json fragments
- parquet stores comples as string
please check it out. maybe if we have a string in jsonl it will start to work

I'd disable JSON support by default and make it configurable. still jsonl keeping complex as dict/list (not string) may be a problem

Reply:
I have disabled it completely for now, and updated the tests. it will be possible to put it in once someone wants to have it. I'm not sure wether it is worth the trouble at this point tbh.